### PR TITLE
feat(file upload): Basic Upload Component

### PR DESCRIPTION
**Pull request by ingriddorioschulze**
This PR implements a basic upload files component **without the functionality**

- user can select files (accepted files are .pdf. .tiff and .tif)
- when upload is done a list of files will be displayed
- for each file, the file name, file size and a delete button will be displayed
- error messages will be displayed 
- when the upload is in progress a loader, the file name, a loading status and a cancel button will be displayed
- when in progress, a button to select more files will be displayed
- a warning message will be displayed when the file limit is reached

### CODE REVIEW

File: app/components/fileUpload/FileUpload.tsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+//   Styling observations:
+//   - Having a custom label attached to the input so I could style it as in the design.
+//   - The input is visually hidden but not hidden from the DOM, so the element can still be there and be accessed.
+//   - I used a span inside a label because I can not give a role to the label and I think it will be not keyboard accessible.
+//   - I am not sure about using tabIndex on the span.
+//   - Using accept=".pdf, .tiff" to restrict the file types that can be uploaded. Is this accessible?
+
+//  Feature needs to clarify:
+//  - Should I already implement a form?
+//  - If yes, I need also to implement a page
```
Comment (Spencer6497) [inline app/components/fileUpload/FileUpload.tsx:18]: luckily I've already done this 😉 if you cross-merge the `basic-file-upload` branch (or rebase your changes on top of it), you should be able to access the `prozesskostenhilfe/formular/file-upload` page (with form)

Comment (Spencer6497) [inline app/components/fileUpload/FileUpload.tsx:18]: luckily I've already done this 😉 if you cross-merge the basic-file-upload branch (or rebase your changes on top of it), you should be able to access the prozesskostenhilfe/formular/file-upload page (with form)

File: app/components/fileUpload/FileUpload.tsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        <div className="w-full h-auto mb-8 mt-8">
+          <Button
+            look="tertiary"
+            iconLeft={<DeleteOutline className="w-6 h-6" />}
+            aria-label="delete uploaded file"
+            text="Entfernen"
+            onClick={() => setFile(null)}
+          />
+        </div>
+      )}
```
Comment (Spencer6497) [inline app/components/fileUpload/FileUpload.tsx:80]: these could be collapsed together into one conditional. Or a ternary, and when `!file` you could display the input/upload button

File: app/components/fileUpload/FileUploadError.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  InvalidFileSize = "InvalidFileSize",
+}
+
+const fileUploadErrorMessage = {
+  [FileUploadErrorType.NoFileUploaded]: "Bitte wählen Sie eine Datei aus.",
+  [FileUploadErrorType.InvalidFileExtension]:
+    "Bitte laden Sie nur PDF– oder TIF–Dateien hoch.",
+  [FileUploadErrorType.InvalidFileSize]:
+    "Bitte laden Sie nur Dateien mit einer maximalen Größe von jeweils 100 MB hoch.",
+};
```
Comment (Spencer6497) [inline app/components/fileUpload/FileUploadError.tsx:24]: will these strings eventually be in stringReplacements instead of living in the code?

File: app/components/fileUpload/FileUploadError.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      return null;
+  }
+};
+
+export const FileUploadError = ({
+  errorMessage,
+  file,
+}: FileUploadErrorProps) => {
+  return (
+    errorMessage && (
```
Comment (Spencer6497) [inline app/components/fileUpload/FileUploadError.tsx:58]: question: should the null check for `errorMessage` come before the return statement with something like this?
```
if (!errorMessage) return null
return { ... }
```
what happens when null is returned in the DOM? i don't actually know 😅

File: app/components/fileUpload/FileUploadError.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    case FileUploadErrorType.NoFileUploaded:
+      return file
+        ? fileUploadErrorMessage[FileUploadErrorType.NoFileUploaded]
+        : null;
+    case FileUploadErrorType.InvalidFileExtension:
+      return !(
+        file?.type === "application/pdf" || file?.type === "application/tiff"
+      )
+        ? null
+        : fileUploadErrorMessage[FileUploadErrorType.InvalidFileExtension];
```
Comment (Spencer6497) [inline app/components/fileUpload/FileUploadError.tsx:43]: i think this condition is inverted

File: app/components/fileUpload/FileUploadError.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    case FileUploadErrorType.InvalidFileSize:
+      return fileBytesToMegabytes <= fileLimitMegabytes
```
Comment (Spencer6497) [inline app/components/fileUpload/FileUploadError.tsx:45]: also here

File: app/components/fileUpload/__test__/FileUpload.test.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        fileName={"testFile.pdf"}
+        fileExtension={"application/pdf"}
+        fileSize={0}
+        state={FileUploadState.NotStarted}
+        errorMessage={FileUploadErrorType.NoFileUploaded}
+      />,
+    );
+
+    const fileUpload = screen.getByRole("button", { name: "Datei auswählen" });
+    expect(fileUpload).toBeInTheDocument();
```
Comment (Spencer6497) [inline app/components/fileUpload/__test__/FileUpload.test.tsx:20]: if we could somehow find a way to target the input component, I think it'd be really nice to [test an upload](https://testing-library.com/docs/user-event/utility#upload)

**Review by Spencer6497 — COMMENTED** [review]
great work! Left a few comments for clarification but i'm excited to see it come to life!

File: app/components/fileUpload/FileUploadError.tsx (hunk 3 above)
Comment (ingriddorioschulze) [reply app/components/fileUpload/FileUploadError.tsx]: Yes, I will move them soon

**Comment by aaschlote** [comment]
> @ingriddorioschulze is ready for review or are you still working on that?
> 
> @aaschlote I am still working 🚧

File: app/components/filesUpload/FilesUpload.tsx [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+import { FilesUploadHeader } from "./FilesUploadHeader";
+import { FilesUploadInProgress } from "./FilesUploadInProgress";
+import { FilesUploadInput } from "./FilesUploadInput";
+import { FilesUploadWarning } from "./FilesUploadWarning";
+
+export enum FilesUploadState {
+  NotStarted = "notStarted",
+  InProgress = "inProgress",
+  Done = "done",
+  Disabled = "disabled",
```
Comment (khallad2) [inline app/components/filesUpload/FilesUpload.tsx:13]: Good Work, just warning.
Disabled = "disabled" is not used.

File: app/components/filesUpload/FilesUpload.tsx [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  warningDescription,
+  uploadProgressLabel,
+  selectFilesButtonLabel,
+  selectMoreFilesButtonLabel,
+}) => {
+  return (
+    <div className="w-full bg-white p-16">
+      <FilesUploadHeader title={title} description={description} />
+
+      {FilesUploadState.NotStarted && (
```
Comment (khallad2) [inline app/components/filesUpload/FilesUpload.tsx:53]: just wandering if the following scenario could happen? 

Since these enum values are strings ("notStarted", "inProgress", "done"... ), they are always truthy. 
This means that every block guarded by such an expression
`{FilesUploadState.NotStarted &
` will always render regardless of any actual state.

if this is the case we can Introduce a dynamic state or prop (for example, uploadState: FilesUploadState)
```
{uploadState === FilesUploadState.NotStarted && (
  <>
    <FilesUploadInput
      selectFilesButtonLabel={selectFilesButtonLabel}
      fieldName={fieldName}
    />
    {uploadState === FilesUploadState.Error && (
      <FilesUploadError errorMessage={errorMessage} />
    )}
  </>
)}
```

File: app/components/filesUpload/__test__/convertFileSize.test.tsx [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import { convertFileSize } from "../convertFileSize";
+
+describe("convertFileSize", () => {
```
Comment (pgurusinga) [inline app/components/filesUpload/__test__/convertFileSize.test.tsx:3]: can we improve the test cases here for the edge cases? for example
- checking whether `524288` is around `0,5 MB`
- what happens if 0
- what happens if `1073741824` will show `1.024 MB`

File: app/components/filesUpload/FilesUpload.tsx (hunk 8 above)
Comment (ingriddorioschulze) [reply app/components/filesUpload/FilesUpload.tsx:13]: Yes, thanks! Good catch :)

File: app/components/filesUpload/FilesUpload.tsx (hunk 9 above)
Comment (ingriddorioschulze) [reply app/components/filesUpload/FilesUpload.tsx:53]: Yes, this scenario can happen, you are right.
In my previous version I did took care of the states, but when I was chatting with Cris, I realized that the states would not be part of this task. This part will come further in the functionality task.

File: app/components/filesUpload/__test__/convertFileSize.test.tsx (hunk 10 above)
Comment (ingriddorioschulze) [reply app/components/filesUpload/__test__/convertFileSize.test.tsx:3]: Sure! Thanks for pointing this out :)

File: app/components/filesUpload/FilesUploadInput.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  selectFilesButtonLabel,
+}) => {
+  return (
+    <div className="w-full my-14">
+      <label>
+        <input
+          data-testid="filesUpload"
+          type="file"
+          name={fieldName}
+          accept=".pdf, .tiff, .tif"
```
[the remark this answers is not in the document]
Comment (ingriddorioschulze) [reply app/components/filesUpload/FilesUploadInput.tsx:20]: That's a really good question! I will ask around and will let you know :)

File: app/components/filesUpload/FilesUpload.tsx [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+  fileSizes: number[];
+  description?: string;
+  warningTitle: string;
+  errorMessage: string;
+  cancelButtonLabel: string;
+  deleteButtonLabel: string;
+  warningDescription: string;
+  uploadProgressLabel: string;
+  selectFilesButtonLabel: string;
+  selectMoreFilesButtonLabel: string;
```
Comment (Spencer6497) [inline app/components/filesUpload/FilesUpload.tsx:30]: idea/nit: we could potentially collapse these into one prop, an object `labels`

File: app/components/filesUpload/FilesUploadDone.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import Button from "../Button";
+
+export type FilesUploadDoneProps = {
+  fileNames: string[];
+  fileSizes: number[];
+  deleteButtonLabel: string;
+  selectMoreFilesButtonLabel: string;
+};
+
+export const FilesUploadDone: FC<FilesUploadDoneProps> = ({
```
Comment (Spencer6497) [inline app/components/filesUpload/FilesUploadDone.tsx:15]: nit: would rename to something like `FinishedFileUpload`

File: app/components/filesUpload/FilesUploadWarning.tsx
Comment (Spencer6497) [inline app/components/filesUpload/FilesUploadWarning.tsx:1]: wondering if we could combine this component with FileUploadError?

**Review by Spencer6497 — APPROVED** [review]
Excellent work ⭐

**Review by pgurusinga — APPROVED** [review]

File: app/components/filesUpload/FilesUpload.tsx [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+export enum FilesUploadState {
+  NotStarted = "notStarted",
+  InProgress = "inProgress",
+  Done = "done",
+  Error = "error",
+  Warning = "warning",
+}
+
+export type FilesUploadProps = {
```
Comment (aaschlote) [inline app/components/filesUpload/FilesUpload.tsx:17]: do u need to export this type?

File: app/components/filesUpload/__test__/convertFileSize.test.tsx [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
```
Comment (aaschlote) [inline app/components/filesUpload/__test__/convertFileSize.test.tsx:3]: can you add test when the array is empty `[]` and has more than one value `[120102012,12121,12100023]`?

**Review by aaschlote — COMMENTED** [review]
Note sure, if it's part of the ticket, but it's missing a11y (accessibility) implementation in this new component. You can check in the component `Input.tsx` looking in the `aria-` tag attributes how it's implemented.

If it's not needed now, I would say LGTM!

File: app/components/filesUpload/FilesUpload.tsx (hunk 14 above)
Comment (ingriddorioschulze) [reply app/components/filesUpload/FilesUpload.tsx:17]: True, not necessary! Thanks :)

File: app/components/filesUpload/FilesUpload.tsx [hunk 16]
```diff
@@ -0,0 +1,91 @@
+import { FC } from "react";
+import { FilesUploadDone } from "./FilesUploadDone";
+import { FilesUploadError } from "./FilesUploadError";
+import { FilesUploadHeader } from "./FilesUploadHeader";
+import { FilesUploadInProgress } from "./FilesUploadInProgress";
+import { FilesUploadInput } from "./FilesUploadInput";
+import { FilesUploadWarning } from "./FilesUploadWarning";
+
+export enum FilesUploadState {
```
[the remark this answers is not in the document]
Comment (ingriddorioschulze) [reply app/components/filesUpload/FilesUpload.tsx:9]: It was in another file and the copy + paste tricked me 😄 
But, since I am using it to render the components and the states will be implemented later, the export is (somehow) keeping me from getting "This condition will always return 'true'" error. Is is ok if I keep it for now until we implement the states?

**Comment by ingriddorioschulze** [comment]
> Note sure, if it's part of the ticket, but it's missing a11y (accessibility) implementation in this new component. You can check in the component `Input.tsx` looking in the `aria-` tag attributes how it's implemented.
> 
> If it's not needed now, I would say LGTM!

Yes, a11y is part of this ticket! Thanks for pointing out, I will take a look at the other Input and implement it :)

File: app/components/filesUpload/__test__/convertFileSize.test.tsx (hunk 15 above)
Comment (ingriddorioschulze) [reply app/components/filesUpload/__test__/convertFileSize.test.tsx:3]: It will not be necessary. Thanks to your observation I realized a tiny bug in the function. The fileSize should not be an `[]` and instead a `number`. The component will handle the file name and size individually. I added the assertion to handle when file size is `null` and `undefined`.
